Add UI for deleting workspaces

waxlamp opened the pull request:
Depends on multinet-app/multinet-server#212 and multinet-app/multinet-server#216 
Closes multinet-app/multinet-server#78 

This PR adds functionality to the "delete workspaces" button.

It adds a confirmation dialog with a built-in delay of 2 seconds before the user can click the "confirm" button.

One issue is that the `GET /workspaces` endpoint doesn't seem to immediately update after the `DELETE /workspaces/<id>` call succeeds. To get around this, I have a one-second delay before the Sidebar component refreshes the workspaces list. One second works but is not guaranteed formally. A better solution might be to manually update the Sidebar component's workspace list (as we do when creating a new workspace), in which case the delay won't matter.

Another important task here is to review just the changes to the `multinet` directory, so that I can cut a release of that and make use of it in the client.

I'd like a bunch of reviews on this for various aspects:
- @JackWilb: please let me know if this is the right interaction for what we need in the Multinet project
- @jtomeck: please confirm that my use of Vuetify is good and correct (and that the style is reasonable)
- @subdavis: I can't formally request a review from you until you accept the invitation to the project, but I'd like to know if my Vue/TypeScript usage here is good

Thanks everyone!

TODO:
- [x] Release multinetjs 0.3.0 (which includes `deleteWorkspace()` function)
- [x] Update client to use multinetjs 0.3.0

waxlamp committed:
Remove fade transition from checkboxes and other controls

waxlamp committed:
Add a deleteWorkspace() method to multinetjs API

waxlamp committed:
Add a delete workspaces dialog

waxlamp committed:
Clean up after the deletion operation

waxlamp committed:
Fix lint errors

waxlamp committed:
Fix typescript errors

JackWilb commented:
Just about to pull it down and test it but I also noticed that, f7c69d5, comes from the other PR for the `no-checkbox-animation` branch. We should merge this branch after that one probably, right?

waxlamp commented:
> Just about to pull it down and test it but I also noticed that, [f7c69d5](https://github.com/multinet-app/multinet/commit/f7c69d5505bcebd597b091c29f304ac5a74ce3be), comes from the other PR for the `no-checkbox-animation` branch. We should merge this branch after that one probably, right?

Yes, I forgot to mention the dependency in the PR description; will add that now.

waxlamp committed:
Bump version number

subdavis reviewed (dismissed):
Seems fine.  I haven't done a lot of typescript yet so I don't feel like I have a great eye for what is and isn't good practice.

Is void the assumed return type of methods where return value is undocumented?  If you return something, is that a compiler warning or does it work?

waxlamp commented:
> Seems fine. I haven't done a lot of typescript yet so I don't feel like I have a great eye for what is and isn't good practice.
> 

What do you think about the artificial delay before retrieving the workspaces again? Would I be better off just updating the local data from the known changes to it, rather than waiting an arbitrary time and asking the server directly?

> Is void the assumed return type of methods where return value is undocumented? If you return something, is that a compiler warning or does it work?

I think so... I think the idea is that if you add a return statement, the compiler will infer the return type if it can, and then warn if something else uses that return value in a type-incompatible way. But of course it's better to just be explicit.

waxlamp committed:
Bump multinet dependency version

multinet@0.3.0 includes the deleteWorkspace() function

JackWilb commented on client/src/components/DeleteWorkspaceDialog.vue:
```diff
@@ -0,0 +1,125 @@
+      } = this;
+
+      selection.forEach(async (ws) => {
+        await api.deleteWorkspace(ws);
```
Shouldn't this be taking a path as the argument, so `workspaces/${ws}`

waxlamp replied:
Since this is the client library, there's no need to force the caller to prepend the "workspaces/" part; if you look at the library source code, you'll see that the function takes care of preparing the correct API path to use to make the DELETE call.

JackWilb replied:
Ah, makes sense!

JackWilb reviewed (changes requested):
Left a note and I'm also getting a bunch of errors when deleting, and the workspace remains.

Here are some of the errors:
```
TypeError: _api__WEBPACK_IMPORTED_MODULE_4__.default.deleteWorkspace is not a function

uncaught exception: Object

[Vue warn]: Invalid prop: type check failed for prop "selection". Expected Object, got Array 

found in

---> <DeleteWorkspaceDialog> at src/components/DeleteWorkspaceDialog.vue
       <VSubheader>
         <VList>
           <VNavigationDrawer>
             <Sidebar> at src/components/Sidebar.vue
               <VApp>
                 <App> at src/App.vue
                   <Root>

[Vue warn]: Invalid prop: type check failed for prop "somethingChecked". Expected Object, got Boolean with value false.

found in

---> <DeleteWorkspaceDialog> at src/components/DeleteWorkspaceDialog.vue
       <VSubheader>
         <VList>
           <VNavigationDrawer>
             <Sidebar> at src/components/Sidebar.vue
               <VApp>
                 <App> at src/App.vue
                   <Root>
```

jtomeck commented on client/src/components/DeleteWorkspaceDialog.vue:
```diff
@@ -0,0 +1,125 @@
+
+    <v-card>
+      <v-card-title
+        class="headline pb-0 pt-3"
```
`class="headline pb-0 pt-3 px-5"`

waxlamp replied:
5d744e22f20ef62335af0971663f0948c15080a3

jtomeck commented on client/src/components/DeleteWorkspaceDialog.vue:
```diff
@@ -0,0 +1,125 @@
+        Delete Workspaces
+      </v-card-title>
+
+      <v-card-text class="px-4 pt-4 pb-1">
```
`class="px-5 py-4"`

waxlamp replied:
5d744e22f20ef62335af0971663f0948c15080a3

jtomeck commented on client/src/components/DeleteWorkspaceDialog.vue:
```diff
@@ -0,0 +1,125 @@
+          depressed
+          color="error"
+          @click="execute"
+          :disabled="disabled"
```
For some reason the button looks disabled initially, but after a couple seconds turns red. Is this delay intentional?

JackWilb replied:
It was:
> One issue is that the GET /workspaces endpoint doesn't seem to immediately update after the DELETE /workspaces/<id> call succeeds. To get around this, I have a one-second delay before the Sidebar component refreshes the workspaces list. One second works but is not guaranteed formally.

waxlamp replied:
Yeah, it's my way of slowing the user down before they do something irreversible.

It's quite likely bad UX. The better solution is to not actually delete, but simply "archive" the workspaces so they no longer show up, but are not actually deleted. See #214.

waxlamp committed:
Use more precise types for props

The JS native type is used at runtime by Vue for typechecking the props,
while the typescript "cast" (via the `as` operator) is used at compile
time.

waxlamp commented:
> Left a note and I'm also getting a bunch of errors when deleting, and the workspace remains.
> 
> Here are some of the errors:
> 
> ```
> TypeError: _api__WEBPACK_IMPORTED_MODULE_4__.default.deleteWorkspace is not a function
> 
> uncaught exception: Object
> 
> [Vue warn]: Invalid prop: type check failed for prop "selection". Expected Object, got Array 
> 
> found in
> 
> ---> <DeleteWorkspaceDialog> at src/components/DeleteWorkspaceDialog.vue
>        <VSubheader>
>          <VList>
>            <VNavigationDrawer>
>              <Sidebar> at src/components/Sidebar.vue
>                <VApp>
>                  <App> at src/App.vue
>                    <Root>
> 
> [Vue warn]: Invalid prop: type check failed for prop "somethingChecked". Expected Object, got Boolean with value false.
> 
> found in
> 
> ---> <DeleteWorkspaceDialog> at src/components/DeleteWorkspaceDialog.vue
>        <VSubheader>
>          <VList>
>            <VNavigationDrawer>
>              <Sidebar> at src/components/Sidebar.vue
>                <VApp>
>                  <App> at src/App.vue
>                    <Root>
> ```

Just for posterity: the webpack error was appearing because the local build was still trying to use an old version of the multinet library, and the prop types error was fixed in b341d35.
